subprojects/core/src/main/java/org/gradle/api/internal/file/collections/jdk7/Jdk7DirectoryWalker.java

```diff
@@ -91,17 +92,30 @@ private FileVisitDetails getFileVisitDetails(Path file, BasicFileAttributes attr
                     File child = file.toFile();
                     FileVisitDetails dirDetails = directoryDetailsHolder.peek();
                     RelativePath childPath = dirDetails != null ? dirDetails.getRelativePath().append(!isDirectory, child.getName()) : rootPath;
-                    return new DefaultFileVisitDetails(child, childPath, stopFlag, fileSystem, fileSystem, isDirectory, attrs.lastModifiedTime().toMillis(), attrs.size());
+                    if (attrs == null) {
```

Comment:
This is a deep, deep change that affects a lot of things, so we'll need a more test coverage to verify that it does not break the other things that use this code (which is basically _everything_ in Gradle).

At the very least, I'd like to see some coverage for what happens with incremental build when a task's inputs or outputs cannot be read, checking that Gradle behaves reasonably in this situation and that the user receives a nice error message that explains what went wrong.

We should also add some coverage for what build logic sees when visiting the elements of a `FileTree`, as we've introduced some breaking changes here.

Reply:
I agree on more test coverage. 

Here the potential breaking change is "some code which will throw `GradleException` before now doesn't throw the exception any more because these files are excluded". For people whose file tree walk won't throw `IOException`s, this has no impact on them. However, I'll try to add some tests as you said.